File: apps/api/app/schemas/extractions.py
"""Schemas for AI extraction and review APIs."""

from __future__ import annotations

from datetime import date, datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.schemas.ai_settings import AiProvider


class ExtractionStatus(StrEnum):
    PROCESSING = "processing"
    READY_FOR_REVIEW = "ready_for_review"
    FAILED = "failed"
    CONFIRMED = "confirmed"
    DISCARDED = "discarded"


class FailureReason(StrEnum):
    INVALID_KEY = "invalid_key"
    RATE_LIMITED = "rate_limited"
    TIMEOUT = "timeout"
    UNREADABLE_FILE = "unreadable_file"
    MALFORMED_RESPONSE = "malformed_response"
    PROVIDER_ERROR = "provider_error"


class ExtractionDraft(BaseModel):
    amount_minor: int | None = None
    extracted_currency: str | None = None
    occurred_on: date | None = None
    vendor_name: str | None = None
    suggested_category: str | None = None

    model_config = ConfigDict(from_attributes=True)


class ExtractionRead(BaseModel):
    id: UUID
    workspace_id: UUID
    file_id: UUID
    provider: AiProvider
    status: ExtractionStatus
    draft: ExtractionDraft | None = None
    failure_reason: FailureReason | None = None
    triggered_by: UUID
    triggered_at: datetime
    confirmed_by: UUID | None = None
    confirmed_at: datetime | None = None
    discarded_by: UUID | None = None
    discarded_at: datetime | None = None
    expense_id: UUID | None = None
    can_edit: bool
    can_discard: bool

    model_config = ConfigDict(from_attributes=True)


class ConfirmExtractionRequest(BaseModel):
    amount_minor: int
    occurred_on: date
    category_id: UUID | None = None
    merchant_name: str | None = None
    description: str | None = None

    model_config = ConfigDict(extra="forbid")

File: apps/web/lib/api/extractions.ts
import { apiFetch } from "./client";
import type { AiProvider } from "./ai-settings";

export type ExtractionStatus =
  | "processing"
  | "ready_for_review"
  | "failed"
  | "confirmed"
  | "discarded";

export type FailureReason =
  | "invalid_key"
  | "rate_limited"
  | "timeout"
  | "unreadable_file"
  | "malformed_response"
  | "provider_error";

export type ExtractionDraft = {
  amount_minor: number | null;
  extracted_currency: string | null;
  occurred_on: string | null;
  vendor_name: string | null;
  suggested_category: string | null;
  suggested_category_id: string | null;
};

export type ExtractionRecord = {
  id: string;
  workspace_id: string;
  file_id: string;
  provider: AiProvider;
  status: ExtractionStatus;
  draft: ExtractionDraft | null;
  failure_reason: FailureReason | null;
  triggered_by: string;
  triggered_at: string;
  confirmed_by: string | null;
  confirmed_at: string | null;
  discarded_by: string | null;
  discarded_at: string | null;
  expense_id: string | null;
  can_edit: boolean;
  can_discard: boolean;
};

export type ConfirmExtractionInput = {
  amountMinor: number;
  occurredOn: string;
  categoryId?: string | null;
  merchantName?: string | null;
  description?: string | null;
};

export async function triggerExtraction(workspaceId: string, fileId: string) {
  return apiFetch<ExtractionRecord>(`/workspaces/${workspaceId}/files/${fileId}/extractions`, {
    method: "POST",
  });
}

export async function listExtractions(workspaceId: string, status?: ExtractionStatus) {
  const query = status ? `?status=${encodeURIComponent(status)}` : "";
  return apiFetch<ExtractionRecord[]>(`/workspaces/${workspaceId}/extractions${query}`);
}

export async function getExtraction(workspaceId: string, extractionId: string) {
  return apiFetch<ExtractionRecord>(`/workspaces/${workspaceId}/extractions/${extractionId}`);
}

export async function confirmExtraction(
  workspaceId: string,
  extractionId: string,
  input: ConfirmExtractionInput,
) {
  return apiFetch<ExtractionRecord>(
    `/workspaces/${workspaceId}/extractions/${extractionId}/confirm`,
    {
      method: "POST",
      body: JSON.stringify({
        amount_minor: input.amountMinor,
        occurred_on: input.occurredOn,
        category_id: input.categoryId ?? null,
        merchant_name: input.merchantName ?? null,
        description: input.description ?? null,
      }),
    },
  );
}

export async function discardExtraction(workspaceId: string, extractionId: string) {
  return apiFetch<ExtractionRecord>(
    `/workspaces/${workspaceId}/extractions/${extractionId}/discard`,
    { method: "POST" },
  );
}
